lib/improver/spotdata_new/neighbour_finding.py

```diff
@@ -0,0 +1,488 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Neighbour finding for the Improver site specific process chain."""
+
+import warnings
+import numpy as np
+from scipy import spatial
+
+import cartopy.crs as ccrs
+
+from improver.utilities.cube_manipulation import enforce_coordinate_ordering
+from improver.spotdata_new.build_spotdata_cube import build_spotdata_cube
+
+
+class NeighbourSelection(object):
+    """
+    For the selection of a grid point near an arbitrary coordinate, where the
+    selection may be the nearest point, or a point that fulfils other
+    imposed constraints.
+
+    Constraints available for determining the neighbours are:
+
+    1. land_constraint which requires the selected point to be on land.
+    2. minimum_dz which minimises the vertical displacement between the
+       given coordinate (when an altitude is provided) and the grid point
+       where its altitude is provided by the relevant model or high resolution
+       orography. Note that spot coordinates provided without an altitude are
+       given the altitude of the nearest grid point taken from the orography
+       cube.
+    3. A combination of the above, where the land constraint is primary and out
+       of available land points, the one with the minimal vertical displacement
+       is chosen.
+    """
+
+    def __init__(self, land_constraint=False, minimum_dz=False,
+                 search_radius=1.0E4,
+                 site_coordinate_system=ccrs.PlateCarree(),
+                 site_x_coordinate='longitude', site_y_coordinate='latitude',
+                 grid_metadata_identifier='mosg'):
+        """
+        Args:
+            land_constraint (bool):
+                If True the selected neighbouring grid point must be on land,
+                where this is determined using a land_mask.
+            minimum_dz (bool):
+                If True the selected neighbouring grid point must be chosen to
+                minimise the vertical displacement compared to the site
+                altitude.
+            search_radius (int or float):
+                The radius in units of the provided area grid coordinates (e.g.
+                metres for a Lambert Azimuthal Equal Areas grid defined in
+                metres), or always in metres for a lat/lon grid.
```

Comment:
I was trying to avoid another unit check in case someone wants to use a cube with coordinates in feet, yards or chains :-) I've made this clearer, making it explicitly metres and added a test for the cube's spatial coordinates in case they don't match.